File: test/elixir/hologram/ex_js_consistency/erlang/elixir_aliases_test.exs
defmodule Hologram.ExJsConsistency.Erlang.ElixirAliasesTest do
  @moduledoc """
  IMPORTANT!
  Each Elixir consistency test has a related JavaScript test in test/javascript/erlang/elixir_aliases_test.mjs
  Always update both together.
  """

  use Hologram.Test.BasicCase, async: true

  @moduletag :consistency

  describe "concat/1" do
    test "works with atom segments which are Elixir module aliases" do
      assert :elixir_aliases.concat([Aaa, Bbb, Ccc]) == Aaa.Bbb.Ccc
    end

    test "works with atom segments which are not Elixir module aliases" do
      assert :elixir_aliases.concat([:Aaa, :Bbb, :Ccc]) == Aaa.Bbb.Ccc
    end

    test "works with binary bitstring segments" do
      assert :elixir_aliases.concat(["Aaa", "Bbb", "Ccc"]) == Aaa.Bbb.Ccc
    end

    test "ignores nil segments" do
      assert :elixir_aliases.concat([Aaa, nil, Ccc]) == Aaa.Ccc
    end

    test "removes the first dot character from the segment before joining segments with a dot character" do
      assert :elixir_aliases.concat(["...Aaa", "...Bbb", "...Ccc"]) == :"Elixir...Aaa...Bbb...Ccc"
    end

    test "doesn't prepend 'Elixir' segment if it is already present as the first element" do
      assert :elixir_aliases.concat(["Elixir", Aaa, Bbb, Ccc]) == Aaa.Bbb.Ccc
    end

    test "raises FunctionClauseError if the argument is not a list" do
      assert_error FunctionClauseError,
                   build_function_clause_error_msg(":elixir_aliases.do_concat/2", [:abc, "Elixir"]),
                   fn -> :elixir_aliases.concat(:abc) end
    end

    test "raises FunctionClauseError if a non-binary bitstring segment is present" do
      assert_error FunctionClauseError,
                   build_function_clause_error_msg(":elixir_aliases.do_concat/2", [
                     [<<1::size(2)>>, "Ccc"],
                     "Elixir.Aaa"
                   ]),
                   fn -> :elixir_aliases.concat(["Aaa", <<1::2>>, "Ccc"]) end
    end

    test "raises FunctionClauseError if any non-atom or non-bitstring segments are present" do
      assert_error FunctionClauseError,
                   build_function_clause_error_msg(":elixir_aliases.do_concat/2", [
                     [123, "Ccc"],
                     "Elixir.Aaa"
                   ]),
                   fn ->
                     :elixir_aliases.concat(["Aaa", 123, "Ccc"])
                   end
    end

    test "raises FunctionClauseError if invalid segment is present and the segments contain 'Elixir' as the first segment" do
      assert_error FunctionClauseError,
                   build_function_clause_error_msg(":elixir_aliases.do_concat/2", [
                     [123, "Ccc"],
                     "Elixir.Aaa"
                   ]),
                   fn ->
                     :elixir_aliases.concat(["Elixir", "Aaa", 123, "Ccc"])
                   end
    end

    test "raises FunctionClauseError if invalid segment is present as the first segment" do
      assert_error FunctionClauseError,
                   build_function_clause_error_msg(":elixir_aliases.do_concat/2", [
                     [123, "Ccc"],
                     "Elixir"
                   ]),
                   fn ->
                     :elixir_aliases.concat([123, "Ccc"])
                   end
    end

    test "error frame carries the do_concat args for a non-list argument" do
      segments = wrap_term(:abc)

      top_frame =
        try do
          :elixir_aliases.concat(segments)
        rescue
          _error -> hd(wrap_term(__STACKTRACE__))
        end

      # The server implements this function in Erlang code inside
      # elixir_aliases.erl, so its frame location also carries the
      # Elixir-internal file and line, which the client doesn't mirror.
      assert {:elixir_aliases, :do_concat, [:abc, "Elixir"], location} = wrap_term(top_frame)
      assert location[:error_info] == nil
    end

    test "error frame carries the do_concat args for an invalid segment" do
      segments = wrap_term(["Aaa", "Bbb", 123])

      top_frame =
        try do
          :elixir_aliases.concat(segments)
        rescue
          _error -> hd(wrap_term(__STACKTRACE__))
        end

      # The server implements this function in Erlang code inside
      # elixir_aliases.erl, so its frame location also carries the
      # Elixir-internal file and line, which the client doesn't mirror.
      assert {:elixir_aliases, :do_concat, [[123], "Elixir.Aaa.Bbb"], location} =
               wrap_term(top_frame)

      assert location[:error_info] == nil
    end
  end

  # Note: there's no atom table limitation in the client-side Hologram runtime,
  # safe_concat/1 can simply delegate directly to concat/1 there.
  # Here we're just testing the function is available.
  test "safe_concat/1" do
    assert :elixir_aliases.safe_concat([Date, Range]) == Date.Range
  end
end

File: test/javascript/erlang/elixir_aliases_test.mjs
"use strict";

import {
  assert,
  assertBoxedError,
  defineGlobalErlangAndElixirModules,
} from "../support/helpers.mjs";

import Erlang_Elixir_Aliases from "../../../assets/js/erlang/elixir_aliases.mjs";
import Interpreter from "../../../assets/js/interpreter.mjs";
import Type from "../../../assets/js/type.mjs";

defineGlobalErlangAndElixirModules();

// IMPORTANT!
// Each JavaScript test has a related Elixir consistency test in test/elixir/hologram/ex_js_consistency/erlang/elixir_aliases_test.exs
// Always update both together.

describe("Erlang_Elixir_Aliases", () => {
  describe("concat/1", () => {
    const concat = Erlang_Elixir_Aliases["concat/1"];
    const expectedAlias = Type.alias("Aaa.Bbb.Ccc");

    it("works with atom segments which are Elixir module aliases", () => {
      const segments = Type.list([
        Type.alias("Aaa"),
        Type.alias("Bbb"),
        Type.alias("Ccc"),
      ]);

      const result = concat(segments);

      assert.deepStrictEqual(result, expectedAlias);
    });

    it("works with atom segments which are not Elixir module aliases", () => {
      const segments = Type.list([
        Type.atom("Aaa"),
        Type.atom("Bbb"),
        Type.atom("Ccc"),
      ]);

      const result = concat(segments);

      assert.deepStrictEqual(result, expectedAlias);
    });

    it("works with binary bitstring segments", () => {
      const segments = Type.list([
        Type.bitstring("Aaa"),
        Type.bitstring("Bbb"),
        Type.bitstring("Ccc"),
      ]);

      const result = concat(segments);

      assert.deepStrictEqual(result, expectedAlias);
    });

    it("ignores nil segments", () => {
      const segments = Type.list([
        Type.alias("Aaa"),
        Type.nil(),
        Type.alias("Ccc"),
      ]);

      const result = concat(segments);
      const expectedAlias = Type.alias("Aaa.Ccc");

      assert.deepStrictEqual(result, expectedAlias);
    });

    it("removes the first dot character from the segment before joining segments with a dot character", () => {
      const segments = Type.list([
        Type.bitstring("...Aaa"),
        Type.bitstring("...Bbb"),
        Type.bitstring("...Ccc"),
      ]);

      const result = concat(segments);
      const expectedAlias = Type.alias("..Aaa...Bbb...Ccc");

      assert.deepStrictEqual(result, expectedAlias);
    });

    it("doesn't prepend 'Elixir' segment if it is already present as the first element", () => {
      const segments = Type.list([
        Type.bitstring("Elixir"),
        Type.alias("Aaa"),
        Type.alias("Bbb"),
        Type.alias("Ccc"),
      ]);

      const result = concat(segments);

      assert.deepStrictEqual(result, expectedAlias);
    });

    it("raises FunctionClauseError if the argument is not a list", () => {
      assertBoxedError(
        () => concat(Type.atom("abc")),
        "FunctionClauseError",
        Interpreter.buildFunctionClauseErrorMsg(":elixir_aliases.do_concat/2", [
          Type.atom("abc"),
          Type.bitstring("Elixir"),
        ]),
      );
    });

    it("raises FunctionClauseError if a non-binary bitstring segment is present", () => {
      const bitstringSize2 = Type.bitstring([
        Type.bitstringSegment(Type.integer(1), {
          type: "integer",
          size: Type.integer(2),
        }),
      ]);

      const segments = Type.list([
        Type.bitstring("Aaa"),
        bitstringSize2,
        Type.bitstring("Ccc"),
      ]);

      assertBoxedError(
        () => concat(segments),
        "FunctionClauseError",
        Interpreter.buildFunctionClauseErrorMsg(":elixir_aliases.do_concat/2", [
          Type.list([bitstringSize2, Type.bitstring("Ccc")]),
          Type.bitstring("Elixir.Aaa"),
        ]),
      );
    });

    it("raises FunctionClauseError if any non-atom or non-bitstring segments are present", () => {
      const segments = Type.list([
        Type.bitstring("Aaa"),
        Type.integer(123),
        Type.bitstring("Ccc"),
      ]);

      assertBoxedError(
        () => concat(segments),
        "FunctionClauseError",
        Interpreter.buildFunctionClauseErrorMsg(":elixir_aliases.do_concat/2", [
          Type.list([Type.integer(123), Type.bitstring("Ccc")]),
          Type.bitstring("Elixir.Aaa"),
        ]),
      );
    });

    it("raises FunctionClauseError if invalid segment is present and the segments contain 'Elixir' as the first segment", () => {
      const segments = Type.list([
        Type.bitstring("Elixir"),
        Type.bitstring("Aaa"),
        Type.integer(123),
        Type.bitstring("Ccc"),
      ]);

      assertBoxedError(
        () => concat(segments),
        "FunctionClauseError",
        Interpreter.buildFunctionClauseErrorMsg(":elixir_aliases.do_concat/2", [
          Type.list([Type.integer(123), Type.bitstring("Ccc")]),
          Type.bitstring("Elixir.Aaa"),
        ]),
      );
    });

    it("raises FunctionClauseError if invalid segment is present as the first segment", () => {
      const segments = Type.list([Type.integer(123), Type.bitstring("Ccc")]);

      assertBoxedError(
        () => concat(segments),
        "FunctionClauseError",
        Interpreter.buildFunctionClauseErrorMsg(":elixir_aliases.do_concat/2", [
          Type.list([Type.integer(123), Type.bitstring("Ccc")]),
          Type.bitstring("Elixir"),
        ]),
      );
    });
  });
});
